rust/consensus/src/lib.rs

```diff
@@ -1,404 +1,290 @@
+        keys.sort();
+        let leader = keys[coin as usize % self.committee.size()];
+
+        // Return its certificate and the certificate's digest.
+        dag.get(&round).map(|x| x.get(&leader)).flatten()
     }
 
-    pub fn compute_support(&self, round: RoundNumber, leader: NodeID) -> Stake {
-        //first check that you see the leaders block
-        if let Some((leader_digest, _)) = self.get_header_digest_by_node_round(leader, round - 4) {
-            //Then get all its direct ancestors
-            if let Some(headers) = self.get_headers(round - 3) {
-                let supporting_digests: Vec<Digest> = headers
-                    .iter()
-                    .filter(|(_, header)| header.has_parent(leader_digest))
-                    .map(|(digest, _)| digest)
-                    .cloned()
-                    .collect();
-
-                // Then all the grandkids
-                if let Some(headers) = self.get_headers(round - 2) {
-                    //for each header we check if it has any parent in supporting_headers
-                    let support_supporters_digest: Vec<Digest> = headers
-                        .iter()
-                        .filter(|(_, header)| header.has_at_least_one_parent(&supporting_digests))
-                        .map(|(digest, _)| digest)
-                        .cloned()
-                        .collect();
-                    // Once more for common_core
-                    if let Some(headers) = self.get_headers(round - 1) {
-                        //for each header we check if it has any parent in supporting_headers
-                        let support_stake: Stake = headers
-                            .iter()
-                            .filter(|(_, header)| {
-                                header.has_at_least_one_parent(&support_supporters_digest)
-                            })
-                            .map(|(_, header)| self.committee.stake(&header.author))
-                            .sum();
-                        //this is the support
-                        return support_stake;
-                    }
-                }
+    /// Order the past leaders that we didn't already commit.
+    fn order_leaders(&self, leader: &Certificate, state: &State) -> Vec<Certificate> {
+        let last_committed_round = state.last_committed.values().max().unwrap();
+        let mut to_commit = vec![leader.clone()];
+        let mut leader = leader;
+        for r in (last_committed_round + 2..leader.round()).rev().step_by(2) {
+            // Get the certificate proposed by the previous leader.
+            let (_, prev_leader) = match self.leader(r, &state.dag) {
+                Some(x) => x,
+                None => continue,
+            };
+
+            // Check whether there is a path between the last two leaders.
+            if self.linked(leader, prev_leader, &state.dag) {
+                to_commit.push(prev_leader.clone());
```

Comment:
any reason you push and reverse instead of push_front and pop?

Reply:
Pop requires a mutable borrow of ordered_leaders, which means we cannot make an other one later (to update the state)